jsbits/isomorphic.js

```diff
@@ -1,20 +1,47 @@
-function copyDOMIntoVTree (vtree) {
-    walk (vtree, document.body.firstChild);
+function copyDOMIntoVTree(vtree) {
+  walk(vtree, document.body.firstChild);
 }
 
-function walk (vtree, node) {
-    var i = 0, vdomChild, domChild;
-    vtree.domRef = node;
-    while (i < vtree.children.length) {
-      vdomChild = vtree.children[i];
-      domChild = node.childNodes[i];
-      if (vdomChild.type === "vtext") {
-	  vdomChild.domRef = domChild;
-	  i++;
-	  continue;
+function walk(vtree, node) {
+  // This is slightly more complicated than one might expect since
+  // browsers will collapse consecutive text nodes into a single text node.
+  // There can thus be fewer DOM nodes than VDOM nodes.
+  var i = 0,
+    j = 0,
+    // This indicates if we are in a block of consecutive text nodes.
+    inTextBlock = false,
+    vdomChild,
+    domChild;
+  vtree.domRef = node;
+  for (i = 0; i < vtree.children.length; i++) {
+    vdomChild = vtree.children[i];
+    domChild = node.childNodes[j];
+    if (vdomChild.type === "vtext") {
+      if (!inTextBlock) {
+        // This is the first text node in a block of consecutive text nodes.
+        inTextBlock = true;
+        if (domChild.nodeType !== Node.TEXT_NODE) {
+          console.error("miso error: Expected text node but found got element");
```

Comment:
just printing a message here doesn't seem beneficial; if it's an error shouldn't we stop?

Reply:
Good point, fixed!